### PR TITLE
Move away from lineWarningLength into validationProvider

### LINKED ISSUE #21144: Git commit box does not warn long messages
Testing #20999 

There is a setting to enable warning for long commit messages. Even though it is enabled, there is no warning shown for long messages. Also it would be helpful to update the description of the setting about the length of the message

### PR DESCRIPTION
**Pull request by joaomoreno**
As discussed, here's another approach to fix https://github.com/Microsoft/vscode/issues/21144

This solution exposes a `validationProvider` on the SCM input, instead of relying on a very specific `lineWarningLength` property. It is a more general solution to the problem, which feels better on the API side of things.

There are a few drawbacks, though:

- It requires renderer <-> ext host communication while there is interaction with the scm input (though debounced).
- The overall feeling isn't great. The UI seems slow, given the debouncing.
- Opens up quite a bit more API.

@jrieken Please review the approach in the API and, given the drawbacks, let me know if this is still your preferred approach.

### CODE REVIEW

File: src/vs/vscode.d.ts [hunk 1]
```diff
@@ -5986,6 +5986,56 @@ declare module 'vscode' {
...
+
+		/**
+		 * A function that will be called to validate input and give a hint to the user.
+		 *
+		 * @param value The current value of the input box.
+		 * @param cursorPosition The cusror position within the input box.
+		 * @return A human readable string which is presented as diagnostic message.
+		 * Return `undefined`, `null`, or the empty string when 'value' is valid.
+		 */
+		validateInput(value: string, cursorPosition: number): ProviderResult<SourceControlInputBoxValidation | undefined | null>;
```
Comment (jrieken) [inline src/vs/vscode.d.ts]: Providers usually have a provide method, esp. when returning a `ProviderResult`. And you also wanna take a look at that type to learn why you don't need `undefined | null` here...

File: src/vs/vscode.d.ts [hunk 2]
```diff
@@ -6002,9 +6052,10 @@ declare module 'vscode' {
 		placeholder: string;
 
 		/**
-		 * The warning threshold for lines in the input box.
+		 * A validation provider for the input box. It's possible to change
+		 * the validation provider simply by setting this property to a different value.
 		 */
-		lineWarningLength: number | undefined;
+		validationProvider: SourceControlInputBoxValidationProvider;
```
Comment (jrieken) [inline src/vs/vscode.d.ts]: Yeah, maybe... It doesn't fill well into the provider approach which usually is about registering multiple providers, like `registerInputBoxValidationProvider('git', myProviderInstance)`. I'd say make it a real provider or let be a simple callback.

**Review by jrieken — CHANGES_REQUESTED** [review]

**Comment by jrieken** [comment]
The simplified (maybe over-simpilied?), down-to-earth, variant is this: [`validateInput?(value: string): string | undefined | null | Thenable<string | undefined | null>;`](https://github.com/Microsoft/vscode/blob/master/src/vs/vscode.d.ts#L1698)

**Comment by joaomoreno** [comment]
@jrieken I initially went with that approach and gave up on it given the lack of precedence. It is not so obvious that one could **set** that property.

**Comment by jrieken** [comment]
Yeah, it is JavaScript'ish but yeah, not commonly used in our API... Lets talk on Monday

**Comment by joaomoreno** [comment]
@jrieken Did the changes we talked about:

- settable function
- moved to proposed

**Pull request merged**